tests/samplers_tests/test_nsgaii.py

```diff
@@ -146,6 +149,116 @@ def constraints_func(trial: FrozenTrial) -> Sequence[float]:
+                        constraints_1d, repeat=n_constraint_dims
+                    ):
+                        t1 = _create_frozen_trial(0, vals1, constraints1)
+                        t2 = _create_frozen_trial(1, vals2, constraints2)
+                        assert _constrained_dominates(t1, t2, directions) == _dominates(
+                            t1, t2, directions
+                        )
+
+
+@pytest.mark.parametrize("direction", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
+@pytest.mark.parametrize("n_constraint_dims", [1, 2])
+def test_constrained_dominates_feasible_vs_infeasible(
+    direction: StudyDirection, n_constraint_dims: int
+) -> None:
+
+    vals_1d = [-float("inf"), -1, 1, float("inf")]
+
+    # Check all pairs of trials consisting of these constraint values.
+    constraints_1d_feasible = [-float("inf"), -1, 0]
+    constraints_1d_infeasible = [float("nan"), 2, float("inf")]
+
+    directions = [direction]
+
+    all_infeasible_constraints = set(
+        itertools.product(
+            constraints_1d_feasible + constraints_1d_infeasible, repeat=n_constraint_dims
+        )
+    ) - set(itertools.product(constraints_1d_feasible, repeat=n_constraint_dims))
+
+    for val1 in vals_1d:
+        for constraints1 in itertools.product(constraints_1d_feasible, repeat=n_constraint_dims):
+            for val2 in vals_1d:
+                for constraints2 in all_infeasible_constraints:
+                    t1 = _create_frozen_trial(0, [val1], constraints1)
+                    t2 = _create_frozen_trial(1, [val2], constraints2)
+                    assert _constrained_dominates(t1, t2, directions)
+                    assert not _constrained_dominates(t2, t1, directions)
+
+
+@pytest.mark.parametrize("direction", [StudyDirection.MINIMIZE, StudyDirection.MAXIMIZE])
+def test_constrained_dominates_infeasible_vs_infeasible(direction: StudyDirection) -> None:
+
+    vals_1d = [-float("inf"), -1, 1, float("inf")]
+
+    # These constraint values have zero violations.
+    # TODO(contramundum53): Currently NaN counts as infeasible but has zero violation.
+    # This should be changed.
+    zero_constraint_values = [-float("inf"), -1, 0, float("nan")]
+
+    # Check all pairs of these constraints.
+    constraints_infeasible_sorted = [
+        [[float("nan"), z] for z in zero_constraint_values]
+        + [[z, float("nan")] for z in zero_constraint_values],
+        [[1, z] for z in zero_constraint_values] + [[z, 1] for z in zero_constraint_values],
+        [[2, z] for z in zero_constraint_values]
+        + [[z, 2] for z in zero_constraint_values]
+        + [[1, 1]],
+        [[3, z] for z in zero_constraint_values]
+        + [[z, 3] for z in zero_constraint_values]
+        + [[1, 2], [2, 1]],
+        [[float("inf"), z] for z in zero_constraint_values]
+        + [[z, float("inf")] for z in zero_constraint_values]
+        + [[1, float("inf")], [float("inf"), 1], [float("inf"), float("inf")]],
+    ]
```

Comment:
Sorry, I couldn't understand what you want to do here. Can you explain a little more?

Reply:
`constraints_infeasible_sorted` is the list of groups of constraint values, sorted by violations. All constraints in `constraints_infeasible_sorted[0]` have violation 0, all constraints in `constraints_infeasible_sorted[1]` have violation 1, and so on.
I check that all constraints in the same group are not comparable with each other, and all constraints in different group have domination relationship.

Reply:
Thanks for the explanation. I understood.